### PR TITLE
[core] Introduce privilege system for catalog based on FileSystem

### PR DESCRIPTION
**Pull request by tsreaper**
### Purpose

This PR introduces an identity-based privilege system for catalogs. Catalogs can now be updated into a privileged catalogs, where privileged users can be created and granted privileges to tables, databases or the whole catalog.

### Tests

* `PrivilegeManagerTest`
* `PrivilegeProcedureITCase`

### API and Format

Yes. This PR introduces two new system tables: user and privilege.

User table is the table which stores all user information. The schema of user table is:
* user (string): user name (primary key)
* sha256 (bytes): sha256 of password

Privilege table is the table storing what privileges each user have. Its schema is:
* name (string): user or role name (primary key)
* entity_type (string): user or role (primary key)
* identifier (string): identifier of object (primary key)
* privilege (string): name of privilege (primary key), see `PrivilegeType`

### Documentation

Yes. Document is also added.

### CODE REVIEW

**Comment by JingsongLi** [comment]
Please finish documentation in this PR.

File: paimon-core/src/main/java/org/apache/paimon/privilege/PrivilegedFileStore.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+import org.apache.paimon.utils.FileStorePathFactory;
+import org.apache.paimon.utils.SnapshotManager;
+import org.apache.paimon.utils.TagManager;
+
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+
+/** {@link FileStore} with privilege checks. */
+public class PrivilegedFileStore<T> implements FileStore<T> {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/privilege/PrivilegedFileStore.java:50]: We should not wrap FileStore, all things should be finished in Table interface.

File: paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java [hunk 2]
```diff
@@ -54,32 +61,43 @@
 import static org.apache.paimon.utils.Preconditions.checkArgument;
 
 /** Common implementation of {@link Catalog}. */
-public abstract class AbstractCatalog implements Catalog {
+public abstract class AbstractCatalog implements Catalog, PrivilegedCatalog {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java:64]: Maybe we should wrap a catalog just like `PrevilegedFileStoreTable`.

**Review by JingsongLi — COMMENTED** [review]
Can you describe storage format for privilege system in the PR description?

File: paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java (hunk 2 above)
Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java:64]: Some catalogs may not use FileSystem based Previlege system.
For example, RestCatalog will register users in server not filesystem.

**Comment by zhangjun0x01** [comment]
hi, @tsreaper , will there be plan to integrate  with apache ranger  in the future? like this: https://doris.apache.org/docs/dev/admin-manual/privilege-ldap/ranger/

File: docs/content/maintenance/manage-privileges.md [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+* The `root` user,
+which is identified by the provided root password when enabling the privilege system.
+This user always has all privileges in the catalog.
+* The `anonymous` user.
+This is the default user if no username and password is provided when creating the catalog.
+
+## Enable Privileges
+
+Paimon currently only supports file-based privilege system.
+Only catalogs with `'metastore' = 'filesystem'` (the default value) support such privilege system.
```
Comment (JingsongLi) [inline docs/content/maintenance/manage-privileges.md:97]: Can we support filesystem privilege for hive catalog too?

File: paimon-core/src/main/java/org/apache/paimon/privilege/FileBasedPrivilegeManager.java [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+                    "Privilege system is not enabled in warehouse " + warehouse + ".");
+        } else {
+            return null;
+        }
+    }
+
+    private void createUserTable() {
+        Options options = new Options();
+        options.set(CoreOptions.BUCKET, 1);
+        Path tableRoot = new Path(warehouse, USER_TABLE_DIR);
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/privilege/FileBasedPrivilegeManager.java:399]: Can we just use Catalog API here?

File: paimon-core/src/main/java/org/apache/paimon/privilege/FileBasedPrivilegeManager.java [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+ * <ul>
+ *   <li>name (string): user or role name (primary key)
+ *   <li>entity_type (string): user or role (primary key)
+ *   <li>identifier (string): identifier of object (primary key)
+ *   <li>privilege (string): name of privilege (primary key), see {@link PrivilegeType}
+ * </ul>
+ */
+public class FileBasedPrivilegeManager implements PrivilegeManager {
+
+    private static final String USER_TABLE_DIR = "user.sys";
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/privilege/FileBasedPrivilegeManager.java:84]: Maybe db is privilege, and table is user?

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**